### ISSUE TITLE
Training log

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#9]
```
python train.py --data-dir ../datasets/MS1M_v3/ --gpu 0,1 --strategy ddp --max_epochs 25 --model resnet101 --batch-size 904 --task pretrain --lr 0.2 --momentum 0.9 --weight-decay 5e-4 --warmup 0 --cooldown 0 --arcface-s 64 --arcface-m 0.5 --precision 16 --loss arcface --scheduler poly
```

In 15 epochs we have this
 
![2022-04-10-132842_2080x1018_scrot](https://user-images.githubusercontent.com/6365065/162629599-e299eef2-8a03-43bd-b80d-e88c11d19333.png)

Maybe this was the problem
https://github.com/vitalwarley/research/blob/852fa5442156b841ede936746452c874c57848f5/model.py#L97-L106
https://github.com/vitalwarley/research/blob/852fa5442156b841ede936746452c874c57848f5/model.py#L210-L218

that is, new `self.linear` layer without `BatchNorm1d`.

However, it could be the `--batch-size 904`. Below are the loss curves for `resnet34` and `batch-size=2048` (this bs is for each GPU).

![2022-04-10-133403_2165x1001_scrot](https://user-images.githubusercontent.com/6365065/162629912-1c8b7bc6-4604-464c-978e-d162e9af483e.png)

**Comment by vitalwarley** [comment]
I started a new training experiment with `--accumulate_grad_batches=2` to increase the effective batch size (2 * bs, now) without memory overhead.

**Comment by vitalwarley** [comment]
At first half of the 4th epoch...

![2022-04-10-192654_2261x1215_scrot](https://user-images.githubusercontent.com/6365065/162642404-9b888eaa-3fdb-4cfb-8114-3e5cb2983d9f.png)

After that, I started a new experiment with a BatchNormalization1d layer after the last linear before computing the ArcFaceLoss.

**Comment by vitalwarley** [comment]
I am having problems with ResNet101 (and ResNet50) and it is not `--accumulate_grad_batches=2` because a rerun ResNet34 and it worked OK like last time.

![2022-04-12-130958_1507x818_scrot](https://user-images.githubusercontent.com/6365065/163006575-d080010b-c7e5-4690-81e2-5a61ed381df7.png)

- Em cinza temos resnet101 com `--overfit_batches=100`.
- Em laranja temos resnet101 com `--overfit_batches=1000`
- Em azul temos resnet50 treinando no dataset inteiro.

Em todos os casos `--batch-size=1024` com `--accumulate_grad_batches=2`.

**Comment by vitalwarley** [comment]
https://github.com/vitalwarley/research/blob/61298c23a2cd8a5d5d753619d23b6e9bc142dffa/model.py#L194-L201

Considering [these differences](https://github.com/vitalwarley/research/issues/11#issuecomment-1096935808)

- [ ] Remove ReLU and test
- [ ] Remove ReLU, add Dropout, and test

**Comment by vitalwarley** [comment]
> * [ ]  Remove ReLU and test

Em azul temos uma resnet101 treinada por 20 épocas com `--overfit_batches=500` (~1/4 do dataset, ou ~1.25mi imagens)

![2022-04-12-173112_1555x879_scrot](https://user-images.githubusercontent.com/6365065/163048535-697cc40b-9f9e-4f31-aa2d-9aa917005be1.png)

**Comment by vitalwarley** [comment]
> * [ ]  Remove ReLU, add Dropout, and test

A [implementação original](https://github.com/deepinsight/insightface/blob/17cdeab12a35efcebc2660453a8cbeae96e20950/recognition/arcface_torch/train.py#L58-L60) usa dropout=0.

**Comment by vitalwarley** [comment]
> * [x]  Try these for batchnorm:
> 
> ```python
> nn.init.constant_(self.features.weight, 1.0)
> self.features.weight.requires_grad = False
> ```

In progress...

**Comment by vitalwarley** [comment]
Fixing batchnorm wrt to insightface didn't work. Other possible causes...

- [ ] Batch size? The batch size used in resnet34/50/101 was 2048, but only on 34 converged.
- [ ] Absence of `pooling` at the end of the backbone. In the original implementation, they flat the input.

**Comment by vitalwarley** [comment]
> * [ ]  Batch size? The batch size used in resnet34/50/101 was 2048, but only on 34 converged.

Seems likely that is not it. `--accumulate_grad_batches=3` with `--batch_size=1024` make no difference.

![2022-04-13-081004_2455x1084_scrot](https://user-images.githubusercontent.com/6365065/163167861-20597002-b5c0-4c9f-9953-3546df290635.png)

- version_14 is this one.
- version_13, the previous comment -- batchnorm changes.

**Comment by vitalwarley** [comment]
> * [ ]  Absence of `pooling` at the end of the backbone. In the original implementation, they flat the input.

That was it. After removing  `global_pool` and flattening the layer output instead, it converged.

![2022-04-17-073348_2162x972_scrot](https://user-images.githubusercontent.com/6365065/163710795-a9c353b8-fc2d-443e-94ff-f624acfd433b.png)
![2022-04-17-073321_2170x990_scrot](https://user-images.githubusercontent.com/6365065/163710798-c632ac23-e901-4cc1-95a4-8d8423a85681.png)

- I don't know why it had a nan in the validation loss curve at the 3rd epoch.

**Comment by vitalwarley** [comment]
On my ROG Strix,

```
python train.py --data-dir ../datasets/MS1M_v3/ --devices 1 --max_epochs 25 --model resnet34 --batch-size 512 --accumulate_grad_batches 8 --task pretrain --lr 0.2 --momentum 0.9 --weight-decay 5e-4 --warmup 0 --cooldown 0 --arcface-s 64 --arcface-m 0.5 --precision 16 --loss arcface --scheduler poly --gradient_clip_val 5
```

Note that bs=2048 (8 * 512). It was necessary because the GPU has only 8GB.

![2022-04-19-112648_1605x898_scrot](https://user-images.githubusercontent.com/6365065/164027031-50bcd71e-95b3-4dce-807b-1ded2bb5bbff.png)

Accuracy on LFW didn't get better after the 2nd epoch. The best accuracy was 68% -- better than random, as the dataset has 50% of positive pairs. 

Another problem was the norm being too high. I fixed a problem that decrease the norm for a batch by 5x. Maybe it was that? I don't think so, but let's see.

At last, 

![2022-04-19-114318_1536x824_scrot](https://user-images.githubusercontent.com/6365065/164030612-c6db4346-cd41-45a4-bb60-d747bb860f29.png)

Train accuracy on the full dataset peaked on ~96% and loss continued to decrease (< 10).

**Comment by vitalwarley** [comment]
```
python val.py --data-dir ../datasets/MS1M_v3/ --batch-size 256 --task pretrain --ckpt-path ../training/research/lightning_logs/version_15/checkpoints/epoch=14-step=18780.ckpt --model resnet101
...
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃      Validate metric      ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│          lfw_acc          │    0.6323332786560059     │
│         lfw_norm          │     648.9758911132812     │
└───────────────────────────┴───────────────────────────┘
```

**Comment by vitalwarley** [comment]
I tried to train resnet101 again with more epochs, but something seems to be wrong compared to [last time](https://github.com/vitalwarley/research/issues/9#issuecomment-1100849905).

![2022-04-22-111917_1616x873_scrot](https://user-images.githubusercontent.com/6365065/164733161-b4bdd7bc-c38f-4485-8167-7617ac3f5984.png)

No difference in CLI:

```
python train.py --data-dir ../datasets/MS1M_v3/ --devices 2 --max_epochs 25 --model resnet101 --batch-size 1024 --accumulate_grad_batches=2 --task pretrain --lr 0.2 --momentum 0.9 --weight-decay 5e-4 --warmup 0 --cooldown 0 --arcface-s 64 --arcface-m 0.5 --precision 16 --loss arcface --scheduler poly
```

I will check if the modifications I made to get #14 right (#16) are the cause of this.

**Comment by vitalwarley** [comment]
I am training it again with `--max-epochs 50` just to be sure that the problem isn't the learning rate. Besides that, the only other thing I could think of as the problem was probably introduced in fd0b9bdb6f5b9f2f92c3a25dc40b0d0fb7edcc1a.

**Comment by vitalwarley** [comment]
> I am training it again with `--max-epochs 50` just to be sure that the problem isn't the learning rate. Besides that, the only other thing I could think of as the problem was probably introduced in [fd0b9bd](https://github.com/vitalwarley/research/commit/fd0b9bdb6f5b9f2f92c3a25dc40b0d0fb7edcc1a).

Yep, that seems to be the problem. 

![2022-04-22-153239_2560x1440_scrot](https://user-images.githubusercontent.com/6365065/164773852-f88287f8-a4d1-4667-a7f2-4fc84663aabe.png)

- version_15 -- first convergence with resnet 101.
- version_20 -- the problem, which seems to be from `--max_epochs 25` 
- version_21 -- new convergence when using `--max_epochs 50`.

Insightface's training uses 25 epochs, though. I will work on that with #13 and #15 sometime soon.

```
╚═╡(research) [12:19] λ diff ../training/research/lightning_logs/version_20/hparams.yaml ../training/research/lightning_logs/version_21/hparams.yaml 
48c48
<   max_epochs: 25
---
>   max_epochs: 50
```

**Comment by vitalwarley** [comment]
Working on #18 right now. I noticed a made a [mistake](https://github.com/vitalwarley/research/issues/18#issuecomment-1119193911) with the dataset. My model should have been trained on MS1M_v2 and not the MS1M_v3. Here is the [training log](https://raw.githubusercontent.com/anxiangsir/insightface_arcface_log/master/ms1mv2_r100/training.log). It shows the training params, which are different from the configs I used for MS1M_v3. Also, there is a `margin_list = [1.0, 0.5, 0.0]` that I didn't understand. When I experiment again, if it doesn't converge, I will investigate it.

**Comment by vitalwarley** [comment]
![2022-05-07-080910_2560x1440_scrot](https://user-images.githubusercontent.com/6365065/167251700-b4709e01-2763-4ae9-a513-1da6c6c0b3d3.png)

## Test sets

```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃       DataLoader 1        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│       agedb_30_acc        │    0.9393332600593567     │    0.9393332600593567     │
│       agedb_30_norm       │    23.123327255249023     │    23.123327255249023     │
│        cfp_fp_acc         │    0.8932856321334839     │    0.8932856321334839     │
│        cfp_fp_norm        │     20.00213623046875     │     20.00213623046875     │
└───────────────────────────┴───────────────────────────┴───────────────────────────┘
```

- The metrics will go up as the model improves -- it is in the 15th epoch yet.

## FIW

```
╚═╡(research) [8:29] λ python val.py --data-dir ../datasets/fiw/ --batch-size 256 --task finetune --model resnet101 --weights ../training/research/lightning_logs/version_22/checkpoints/epoch=15-step=22752.ckpt --mining-strategy baseline
# ...
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃      Validate metric      ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│          val_acc          │    0.5001000165939331     │
│          val_auc          │    0.5583887100219727     │
└───────────────────────────┴───────────────────────────┘
```

- No difference from [before](https://github.com/vitalwarley/research/issues/18#issuecomment-1107950233). I will try again after training finishes.

**Comment by vitalwarley** [comment]
I tried with the model from the 22nd epoch, but nothing changed for good.

**Comment by vitalwarley** [comment]
```
Training: 2022-04-27 13:01:41,424-[lfw][112000]XNorm: 21.779004
Training: 2022-04-27 13:01:41,424-[lfw][112000]Accuracy-Flip: 0.99750+-0.00281
Training: 2022-04-27 13:01:41,425-[lfw][112000]Accuracy-Highest: 0.99817
Training: 2022-04-27 13:02:31,736-[cfp_fp][112000]XNorm: 22.111145
Training: 2022-04-27 13:02:31,737-[cfp_fp][112000]Accuracy-Flip: 0.98529+-0.00542
Training: 2022-04-27 13:02:31,737-[cfp_fp][112000]Accuracy-Highest: 0.98629
Training: 2022-04-27 13:03:15,249-[agedb_30][112000]XNorm: 22.232803
Training: 2022-04-27 13:03:15,250-[agedb_30][112000]Accuracy-Flip: 0.98167+-0.00882
Training: 2022-04-27 13:03:15,250-[agedb_30][112000]Accuracy-Highest: 0.98250
```

My results aren't the same for `cfp_fp` and `agedb_30`.
